### ISSUE TITLE
Latest Docker Image Error Python failed to call 'asyncio.get_event_loop'

### ISSUE THREAD

**Issue by treyrich** [nginx/unit#815]
Upgrading our docker image from 1.28.0-python3.10 to 1.29.0-python3.11 results in the following log output:
```
2022/12/20 18:38:19 [info] 21#21 unit 1.29.0 started
2022/12/20 18:38:19 [info] 28#28 discovery started
2022/12/20 18:38:19 [notice] 28#28 module: python 3.11.1 "/usr/lib/unit/modules/python3.unit.so"
2022/12/20 18:38:19 [info] 24#24 controller started
2022/12/20 18:38:19 [notice] 24#24 process 28 exited with code 0
2022/12/20 18:38:19 [info] 33#33 router started
2022/12/20 18:38:19 [info] 33#33 OpenSSL 1.1.1n  15 Mar 2022, 101010ef
{
	"certificates": {},
	"config": {
		"listeners": {},
		"applications": {}
	},

	"status": {
		"connections": {
			"accepted": 0,
			"active": 0,
			"idle": 0,
			"closed": 0
		},

		"requests": {
			"total": 0
		},

		"applications": {}
	}
}
/usr/local/bin/docker-entrypoint.sh: Looking for certificate bundles in /docker-entrypoint.d/...
/usr/local/bin/docker-entrypoint.sh: Looking for configuration snippets in /docker-entrypoint.d/...
/usr/local/bin/docker-entrypoint.sh: Applying configuration /docker-entrypoint.d/config.json
2022/12/20 18:38:19 [info] 44#44 "webapp" prototype started
2022/12/20 18:38:19 [info] 46#46 "webapp" application started
2022/12/20 18:38:19 [alert] 0#46 [unit] Python failed to call 'asyncio.get_event_loop'
2022/12/20 18:38:20 [notice] 44#44 app process 46 exited with code 1
2022/12/20 18:38:20 [warn] 33#33 failed to start application "webapp"
2022/12/20 18:38:20 [alert] 33#33 failed to apply new conf
2022/12/20 18:38:20 [notice] 24#24 process 44 exited with code 0
/usr/local/bin/docker-entrypoint.sh: Error: HTTP response status code is '500'
 }error": "Failed to apply new configuration."
```

I've confirmed that this is a bug by replicating it with the simplest python application (copied from Unit docs):
```python

async def application(scope, receive, send):

    await send({"type": "http.response.start", "status": 200})

    await send({"type": "http.response.body", "body": b"Hello, ASGI\n"})
```

Configuration is:
```json
{
  "listeners": {
    "*:8080": {
      "pass": "applications/webapp"
    }
  },

  "applications": {
    "webapp": {
      "type": "python",
      "path": "/www/",
      "module": "nunittest.asgi",
      "callable": "application"
    }
  }
}
```

There appears to be no path forward to using 1.29.0-python3.11 for an asgi application at all at this point, but perhaps I'm missing something?

**Comment by thresheek** [comment]
Hi @treyrich !

Our internal tests for docker images did not catch that issue, however we are testing an wsgi app there.

We're going to investigate the possible causes.

Many thanks

**Comment by treyrich** [comment]
Great, thanks for the quick reply. Do let me know if there's anything I can do to help.

**Comment by thresheek** [comment]
It also seems to work with Python 3.11 as shipped with Fedora 37 (and where our automated tests were run, which cover asgi also).

**Comment by andrew-nuwber** [comment]
Our team has the same issue running python app on quart

**Comment by ac000** [comment]
Trying to reproduce this locally (current unit master, Python 3.11).

With the above config I get

```
2022/12/21 15:13:58 [alert] 9398#9398 Python failed to import module "nunittest.asgi"
ModuleNotFoundError: No module named 'nunittest'
```

With the following

```JSON
{
  "listeners": {
    "[::1]:8080": {
      "pass": "applications/webapp"
    }
  },

  "applications": {
    "webapp": {
      "type": "python",
      "path": "/srv/unit/python/app/",
      "module": "asgi"
    }
  }
}
```

```$ cat /srv/unit/python/app/asgi.py```
```Python
async def application(scope, receive, send):
    await send({"type": "http.response.start", "status": 200})

    await send({"type": "http.response.body", "body": b"Hello, ASGI\n"})
```

The app seems to start, but not really...

```JSON
{
        "success": "Reconfiguration done."
}
```

From the log

```
2022/12/21 15:28:06 [info] 9642#9642 "webapp" application started
2022/12/21 15:28:06 [warn] 0#9642 [unit] Got invalid state transition on lifespan protocol
2022/12/21 15:28:06 [info] 0#9642 [unit] ASGI Lifespan processing exception
Traceback (most recent call last):
  File "/srv/unit/python/app/asgi.py", line 2, in application
    await send({"type": "http.response.start", "status": 200})
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Got invalid state transition on lifespan protocol
```

And when trying to access it

```shell
$ curl http://localhost:8080/
<!DOCTYPE html><title>Error 503</title><p>Error 503.
```
```
2022/12/21 15:28:45 [alert] 9642#9642 [unit] #8: Python failed to create 'client' pair
2022/12/21 15:28:45 [alert] 9642#9642 [unit] Python failed to call 'loop.call_soon'
ValueError: invalid literal for int() with base 10: ':1'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib64/python3.11/asyncio/base_events.py", line 762, in call_soon
    handle = self._call_soon(callback, args, context)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/asyncio/base_events.py", line 778, in _call_soon
    handle = events.Handle(callback, args, self, context)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SystemError: <class 'asyncio.events.Handle'> returned a result with an exception set
```

**Comment by treyrich** [comment]
Here, perhaps this will help. This is the simplest recreation I can come up with. Create a Dockerfile with the following contents:
```dockerfile
# Uncomment 1.28 to see working version
# FROM public.ecr.aws/nginx/unit:1.28.0-python3.10
FROM public.ecr.aws/nginx/unit:1.29.0-python3.11

WORKDIR /www
# port used by the listener in nginx-unit-config.json
EXPOSE 8080
# Copy our Nginx configuration file to the container
RUN echo '{"listeners":{"*:8080":{"pass":"applications/webapp"}},"applications":{"webapp":{"type":"python 3","path":"/www/","module":"nunittest.asgi"}}}' > /docker-entrypoint.d/config.json
# Copy our application code to the container
RUN mkdir /www/nunittest
RUN echo '' > /www/nunittest/__init__.py
RUN echo 'async def application(scope, receive, send):\n\
    await send({"type": "http.response.start", "status": 200})\n\
    \n\
    await send({"type": "http.response.body", "body": b"Hello, ASGI\\n"})\n' > /www/nunittest/asgi.py
```

Then run it with this command `docker run --publish 8080:8080 --rm $(docker build -q .)` which will build and run the dockerfile in the current directory.

You'll see that with 1.29.0 the server crashes, but if you uncomment the 1.28.0 image and comment out 1.29.0 that the server launches without issue.

**Comment by hawiliali** [comment]
Hi, 
I updated to 1.29.0, I use FastAPI with python 3.10 without issues
Using the sample app in your documentation [https://unit.nginx.org/howto/fastapi/](FastAPI how to) with python 3.11 yielded same error
`
2022/12/26 12:20:07 [info] 8422#8422 "fastapi" prototype started
2022/12/26 12:20:08 [info] 8427#8427 "fastapi" application started
2022/12/26 12:20:08 [alert] 0#8427 [unit] Python failed to call 'asyncio.get_event_loop'
2022/12/26 12:20:09 [notice] 8422#8422 app process 8427 exited with code 1
2022/12/26 12:20:09 [warn] 30882#30882 failed to start application "fastapi"
2022/12/26 12:20:09 [alert] 30882#30882 failed to apply new conf
2022/12/26 12:20:09 [notice] 30879#30879 process 8422 exited with code 0
`

**Comment by hawiliali** [comment]
Hi,

Using the sample fastapi app in your documentation [https://unit.nginx.org/howto/fastapi/]()
After updating to 1.29.0 with python 3.11 , yielded same error in this issue
``
2022/12/26 12:20:07 [info] 8422#8422 "fastapi" prototype started
2022/12/26 12:20:08 [info] 8427#8427 "fastapi" application started
2022/12/26 12:20:08 [alert] 0#8427 [unit] Python failed to call 'asyncio.get_event_loop'
2022/12/26 12:20:09 [notice] 8422#8422 app process 8427 exited with code 1
2022/12/26 12:20:09 [warn] 30882#30882 failed to start application "fastapi"
2022/12/26 12:20:09 [alert] 30882#30882 failed to apply new conf
2022/12/26 12:20:09 [notice] 30879#30879 process 8422 exited with code 0
``
using same sample  application with python 3.10, I have no issues (compiled the module)
looks like the issue with ASGI apps in general using unit 1.29.0 with python 3.11

**Comment by mamashin** [comment]
Same error with python 3.11.1 and unit 1.29.0 :(

**Comment by tippexs** [comment]
Hi Team - I am working with the Dev Team on an analysis about the root cause. Will let you all know once we have something to share.

Atm. it looks like it is something new in the combination of Unit 1.29 and Python 3.11.1 as Python 3.10 on 1.29 works just fine.

**Comment by andrew-nuwber** [comment]
@tippexs hi, thank you for working on that case!

Just like temporary workaround would be great to have docker image with unit 1.29.0 and python 3.10

**Comment by tippexs** [comment]
@here this sounds like a plan. Let me share a Dockerfile that will let you build a Image with Unit1.29 as well as Python 3.10 and a language module.

**Comment by tippexs** [comment]
Or if this would help in the meantime:

http://unit.nginx.org/installation/#inst-lang-docker

```shell
git clone https://github.com/nginx/unit
cd unit
git checkout 1.29.0  # Optional; use to choose a specific Unit version
cd pkg/docker/
make build-python3.10 VERSION_python=3.10
```

**Comment by tippexs** [comment]
I have just tested Unit 1.29 using Python 3.10 and I get the same error as with Python 3.11. Currently we can not built 1.28 with Python 3.11. We are looking into the issue but at the moment it would be better to stay on 1.28 with Python 3.10 till we know whats going on with Unit 1.29 and Python ASGIs.

**Comment by tippexs** [comment]
Okay scratch that - It was an issue on my end! I am now trying to play around with the images.

**Comment by tippexs** [comment]
Updates! I was able to make it work with Unit 1.29 and Python 3.10.7. As soon as I am trying to use the latest Python 3.10 version which is 3.10.9 the app starts to crash on boot.

To build a container with Unit 1.29 and Python 3.10.7 issue
```shell
git clone https://github.com/nginx/unit
cd unit
git checkout 1.29.0  # Optional; use to choose a specific Unit version
cd pkg/docker/
make build-python3.10 VERSION_python=3.10.7                                                                                                                                                                                                         
```

This will create a Docker image tagged as `unit:1.29.0-python3.10`.  Feel free to test your apps with this image. In the meantime we are looking into the issue with Python Versions > 3.10.7

**Comment by 88Ocelot** [comment]
i had managed to run django asgi app with python3.11 using asyncio [runner](https://docs.python.org/3/library/asyncio-runner.html#asyncio.Runner) instead of using [get_event_loop](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_event_loop ) and [new_event_loop](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.new_event_loop). i dont think its the perfect solution, but i havent seen any memory leaks or leaks after applying load testing for an hour. 
#826

**Comment by ac000** [comment]
Hi @88Ocelot. First of all thank you for taking the time and effort to implement a fix for this. However, and I'm sure you'll agree, that it's a rather large patch and fundamentally changes how we do the event loop.

So I would rather propose the following more minimal change

```diff
diff --git a/src/python/nxt_python_asgi.c b/src/python/nxt_python_asgi.c
index 587a17cf..9fe5b56e 100644
--- a/src/python/nxt_python_asgi.c
+++ b/src/python/nxt_python_asgi.c
@@ -201,14 +201,53 @@ nxt_python_asgi_init(nxt_unit_init_t *init, nxt_python_proto_t *proto)
 }
 
 
+static PyObject *
+nxt_python_asgi_get_event_loop(PyObject *asyncio, const char *event_loop_func)
+{
+    PyObject  *event_loop, *loop;
+
+    event_loop = PyDict_GetItemString(PyModule_GetDict(asyncio),
+                                      event_loop_func);
+    if (nxt_slow_path(event_loop == NULL)) {
+        nxt_unit_alert(NULL, "Python failed to get '%s' from module 'asyncio'",
+                       event_loop_func);
+        return NULL;
+    }
+
+    if (nxt_slow_path(PyCallable_Check(event_loop) == 0)) {
+        nxt_unit_alert(NULL, "'asyncio.%s' is not a callable object",
+                       event_loop_func);
+        return NULL;
+    }
+
+    loop = PyObject_CallObject(event_loop, NULL);
+    if (nxt_slow_path(loop == NULL)) {
+        if (strcmp(event_loop_func, "get_running_loop") != 0) {
+            nxt_unit_alert(NULL, "Python failed to call 'asyncio.%s'",
+                           event_loop_func);
+        }
+
+        return NULL;
+    }
+
+    return loop;
+}
+
+
 static int
 nxt_python_asgi_ctx_data_alloc(void **pdata, int main)
 {
     uint32_t                i;
-    PyObject                *asyncio, *loop, *event_loop, *obj;
+    PyObject                *asyncio, *loop, *obj;
     const char              *event_loop_func;
     nxt_py_asgi_ctx_data_t  *ctx_data;
 
+#if PY_VERSION_HEX < NXT_PYTHON_VER(3, 7)
+    static const char       *main_event_loop_func = "get_event_loop";
+#else
+    static const char       *main_event_loop_func = "get_running_loop";
+#endif
+
     ctx_data = nxt_unit_malloc(NULL, sizeof(nxt_py_asgi_ctx_data_t));
     if (nxt_slow_path(ctx_data == NULL)) {
         nxt_unit_alert(NULL, "Failed to allocate context data");
@@ -241,29 +280,25 @@ nxt_python_asgi_ctx_data_alloc(void **pdata, int main)
         goto fail;
     }
 
-    event_loop_func = main ? "get_event_loop" : "new_event_loop";
+    event_loop_func = main ? main_event_loop_func : "new_event_loop";
 
-    event_loop = PyDict_GetItemString(PyModule_GetDict(asyncio),
-                                      event_loop_func);
-    if (nxt_slow_path(event_loop == NULL)) {
-        nxt_unit_alert(NULL,
-                       "Python failed to get '%s' from module 'asyncio'",
-                       event_loop_func);
+    loop = nxt_python_asgi_get_event_loop(asyncio, event_loop_func);
+    if (loop == NULL) {
+#if PY_VERSION_HEX < NXT_PYTHON_VER(3, 7)
         goto fail;
-    }
+#else
+        if (main) {
+            PyErr_Clear();
 
-    if (nxt_slow_path(PyCallable_Check(event_loop) == 0)) {
-        nxt_unit_alert(NULL,
-                       "'asyncio.%s' is not a callable object",
-                       event_loop_func);
-        goto fail;
-    }
+            loop = nxt_python_asgi_get_event_loop(asyncio, "new_event_loop");
+            if (nxt_slow_path(loop == NULL)) {
+                goto fail;
+            }
 
-    loop = PyObject_CallObject(event_loop, NULL);
-    if (nxt_slow_path(loop == NULL)) {
-        nxt_unit_alert(NULL, "Python failed to call 'asyncio.%s'",
-                       event_loop_func);
-        goto fail;
+        } else {
+            goto fail;
+        }
+#endif
     }
 
     for (i = 0; i < nxt_nitems(handlers); i++) {
```

which AFAICT has the same behaviour as previously. So if you are able, please do give this patch a spin. I have tested it against several versions of Python but only as far as ensuring the _asyncio.get_event_loop_ error doesn't show up and that the (very basic) application starts.

**Comment by alejandro-colomar** [comment]
> ```diff
> diff --git a/src/python/nxt_python_asgi.c b/src/python/nxt_python_asgi.c
> index 587a17cf..9fe5b56e 100644
> --- a/src/python/nxt_python_asgi.c
> +++ b/src/python/nxt_python_asgi.c
> ```
[...]

Please, split that into a refactoring patch that adds the function, and then one that does the fix, to keep my sanity when reviewing :)

**Comment by ac000** [comment]
I have pushed a [branch](https://github.com/ac000/unit/tree/asyncio) that contains the above fix, if that facilitates easier testing.

**Comment by luckydonald** [comment]
When will this be merged? 
And if that's not yet planned, what are the current workarounds, now that the problem is [understood](https://github.com/nginx/unit/commit/7dcbd778ac21fe2cc76ea7f1761e41e91a398230)?

I'm using `FROM nginx/unit:1.29.0-python3.11` in my docker file, as I require py3.11.

**Comment by luckydonald** [comment]
I think a workaround is the following for now:

`config`:
```diff
        "backend": {
            "type": "python3",
            <your custom stuff stays as before>
            "protocol": "asgi",
-           "module": "app",
+           "module": "workaround-issue815",
            "callable": "app"
        }
```

`workaround-issue815.py`:
```py
import asyncio

# workaround for [unit] Python failed to call 'asyncio.get_event_loop'
# See https://github.com/nginx/unit/issues/815 for the bug status,
# see also: https://github.com/nginx/unit/commit/7dcbd778ac21fe2cc76ea7f1761e41e91a398230 for an implementation
try:
    asyncio.get_running_loop()
except RuntimeError:
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
# end def

# noinspection PyUnresolvedReferences
from app import app  # basically the old setting: from "module" import "callable"
```

**Comment by ac000** [comment]
@luckydonald Rather than workarounds you could test the proposed fix... that might speed things along...

Either the patch or the linked branch.

**Comment by luckydonald** [comment]
Sure, could help me out with an docker image or Dockerfile?
I'd gladly will test it out.

**Comment by luckydonald** [comment]
@ac000 sure (see message above)

**Comment by rafsaf** [comment]
Hello, thanks for working on it, waiting patiently here for this bug to be fixed, I think everything is already clear.

**Comment by lcrilly** [comment]
> waiting patiently here for this bug to be fixed

@rafsaf we have a fix. Are you in a position to test it?
https://github.com/ac000/unit/tree/asyncio

**Comment by rafsaf** [comment]
@lcrilly sure thing I can do it, hope I can find time today and try it on evening with an app that broke after image upgrade

**Comment by tippexs** [comment]
Hi @rafsaf 
I have just tested the build process with the Patch from @ac000 and to build the Docker Image using the Pkg-Makefile you can simply change the `template.Dockerfile` in `pkg/docker/.

Currently it is (https://github.com/nginx/unit/blob/5a37171f733fa8c7326161cc49af3df0be5dfae4/pkg/docker/template.Dockerfile#L9)
```shell
    && hg clone https://hg.nginx.org/unit \
    && cd unit \
    && hg up @@VERSION@@ \
```
For your tests it should be 
```shell
    && git clone -b asyncio https://github.com/ac000/unit.git \
    && cd unit \
```

This will pull down the correct sources. I am about to make this configurable in the make process to accept `git`, `hg` [default] and `local` sources.

After changing the `template.Dockerfile` you can run

```shell
make build-python3.10 VERSION_python=3.10
```

Let me know if that works.

**Comment by rafsaf** [comment]
@tippexs Awesome, probably there would a bit of headache if not you ;) Not a bad idea to clone repo in Dockerfile btw.

But to the point, I was able to build locally both `unit:1.30.0-python3.10` using `make build-python3.10 VERSION_python=3.10` and `unit:1.30.0-python3.11` using `make build-python3.11 VERSION_python=3.11` just as you described it.

**Both are working alright** with my setup where current official image `nginx/unit:1.29.0-python3.11` fails.

Steps to reproduce results if anybody is concerned (python 3.10/3.11 fastapi asgi app) can be found here https://github.com/rafsaf/nginx_unit_issue_815


If there is anything I could do, don't hesitate, glad I could help.

**Comment by ac000** [comment]
@rafsaf Thanks for taking the time and effort to test this!

I will look at getting the patches reviewed and merged.

**Comment by ac000** [comment]
The fix for this has been merged.

**Comment by micaelmalta** [comment]
Thanks @ac000 ! 

When can we expect an official docker release for this fix?

**Comment by thresheek** [comment]
Next Unit release will have the fixes.

**Comment by micaelmalta** [comment]
Do we have any ETA?

**Comment by lcrilly** [comment]
Hi @micaelmalta 

We're making plans for the next release now. Will update you in the next week or so. In the meantime, we can provide a Dockerfile that will produce a working Unit 1.29 + Python build if that will help you?

**Comment by rafsaf** [comment]
Thanks for eta

It could be more scalable (also for situations like that in the future) for unit to have nightly releases of docker images, something that for example rust does and many others nowadays https://doc.rust-lang.org/book/appendix-07-nightly-rust.html.

This way you wouldn't need to rush with new stable releases or deal with one time images 😀

**Comment by 0x6f677548** [comment]
Just for the sake of others: this seems resolved in docker image 1.29.1-python3.11

**Comment by lcrilly** [comment]
Thanks @0x6f677548 👍 

Sorry I dropped the ball on the comms on this one. It was indeed fixed with [this fix](https://github.com/nginx/unit/commit/9c0a4a09978e0defbc6656c7866a7ca147236733) and shipped with [1.29.1](https://unit.nginx.org/CHANGES.txt).

**Comment by luckydonald** [comment]
Works there without problems.
